titus-server-master/src/main/java/com/netflix/titus/master/service/management/kube/KubeApplicationSlaManagementService.java

```diff
@@ -0,0 +1,185 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.master.service.management.kube;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.annotation.PreDestroy;
+import javax.inject.Singleton;
+
+import com.netflix.titus.api.model.ApplicationSLA;
+import com.netflix.titus.api.service.TitusServiceException;
+import com.netflix.titus.common.runtime.TitusRuntime;
+import com.netflix.titus.common.util.CollectionsExt;
+import com.netflix.titus.common.util.StringExt;
+import com.netflix.titus.common.util.rx.ReactorExt;
+import com.netflix.titus.master.service.management.ApplicationSlaManagementService;
+import com.netflix.titus.master.service.management.kube.crd.F8IOCapacityGroup;
+import com.netflix.titus.master.service.management.kube.crd.F8IOCapacityGroupSpec;
+import com.netflix.titus.master.service.management.kube.crd.Fabric8IOModelConverters;
+import io.fabric8.kubernetes.api.model.KubernetesResourceList;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.NamespacedKubernetesClient;
+import io.fabric8.kubernetes.client.dsl.NonNamespaceOperation;
+import io.fabric8.kubernetes.client.dsl.Resource;
+import io.fabric8.kubernetes.client.informers.ResourceEventHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import reactor.core.scheduler.Schedulers;
+import rx.Observable;
+
+@Singleton
+public class KubeApplicationSlaManagementService implements ApplicationSlaManagementService {
+
+    private static final Logger logger = LoggerFactory.getLogger(KubeApplicationSlaManagementService.class);
+
+    private final CapacityGroupInformerWrapper capacityGroupInformer;
+    private final NonNamespaceOperation<F8IOCapacityGroup, KubernetesResourceList<F8IOCapacityGroup>, Resource<F8IOCapacityGroup>> crdClient;
+    private final Scheduler scheduler;
+
+    private volatile Map<String, ApplicationSLA> capacityGroups = Collections.emptyMap();
+
+    public KubeApplicationSlaManagementService(NamespacedKubernetesClient kubeApiClient, TitusRuntime titusRuntime) {
+        this.crdClient = kubeApiClient.resources(F8IOCapacityGroup.class).inNamespace("default");
+        this.capacityGroupInformer = new CapacityGroupInformerWrapper(kubeApiClient, titusRuntime);
+        this.scheduler = Schedulers.boundedElastic();
+
+        capacityGroupInformer.getInformer().addEventHandler(new ResourceEventHandler<F8IOCapacityGroup>() {
```

Comment:
Have you looked at this?
https://javadoc.io/static/io.kubernetes/client-java/8.0.0/io/kubernetes/client/informer/cache/Cache.html

You shouldn't have to write your own cache.

Reply:
I do not think it supports value transformers? We cache `ApplicationSla` objects to avoid constant mapping from CRDs.